[AOTInductor] Simplified AOTInductor interface and model class

chenyang78 opened the pull request:
Summary:
This PR removed several APIs from the AOTInductor interface,
which are not used by the client.

It also simplified AOTInductor's model class by removing
the dim info for input/output tensors. We included dim info
before to return max output shapes, which was used by the client
to allocate memory for output tensors. Now, we allocate output
tensor memory from the .so so that we don't need to maintain
such information any more. The deletion of dim info from
the model class also simplified the codegen quite a bit.

Test Plan: ci

Reviewed By: khabinov

Differential Revision: D49835430




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

khabinov reviewed (approved):
Thanks, LGTM!

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1254,14 +1220,8 @@ def codegen_model_constructor(self):
         AOTInductorModel::AOTInductorModel()
             : AOTInductorModelBase(4, 1) {
-        auto s0 = make_dynamic_dim("s0", 2048);
         inputs_info_[0].name = "input0";
```
Since you are working on simplification, let me ask this question: do we really need to maintain the name and dtype for inputs and outputs? If we do, then should we make them as members of AOTInductorModelContainer, as they are the same for the model instances?

chenyang78 replied:
I think we'd better keep input/output names, because the order of inputs and outputs kept in the .so might not be the same as the order of inputs and outputs prepared by the user. Having an API that returns the name at an index would allow the user to create correct mappings. We don't really need ```dtype```. Let me remove it. 

Regarding where to keep input/output name info, I think attaching them to the model class may be better, because the user may not use the model container via the new APIs ```AOTInductorModelCreate``` and ```AOTInductorModelRun```.

Commit:
[AOTInductor] Simplified AOTInductor interface and model class (#110411)

Summary:

This PR removed several APIs from the AOTInductor interface,
which are not used by the client.

It also simplified AOTInductor's model class by removing
the dim info for input/output tensors. We included dim info
before to return max output shapes, which was used by the client
to allocate memory for output tensors. Now, we allocate output
tensor memory from the .so so that we don't need to maintain
such information any more. The deletion of dim info from
the model class also simplified the codegen quite a bit.

Test Plan: ci

Reviewed By: khabinov

Differential Revision: D49835430
